File: firmware/src/protocol/action_codec.cpp
#include "protocol/action_codec.hpp"

namespace streamdeck::protocol {
namespace {
uint32_t read_u32(const uint8_t* data) {
    return static_cast<uint32_t>(data[0]) |
           (static_cast<uint32_t>(data[1]) << 8) |
           (static_cast<uint32_t>(data[2]) << 16) |
           (static_cast<uint32_t>(data[3]) << 24);
}

void write_u32(uint8_t* data, uint32_t value) {
    data[0] = static_cast<uint8_t>(value);
    data[1] = static_cast<uint8_t>(value >> 8);
    data[2] = static_cast<uint8_t>(value >> 16);
    data[3] = static_cast<uint8_t>(value >> 24);
}
}

bool is_valid_action(const actions::Action& action) {
    using actions::ActionType;
    switch (action.type) {
    case ActionType::None:
        return action.key_code == 0 && action.modifiers == 0 &&
               action.consumer == actions::ConsumerAction::VolumeUp &&
               action.host_action_id == 0;
    case ActionType::Keyboard:
        return action.key_code != 0 && action.modifiers == 0 &&
               action.consumer == actions::ConsumerAction::VolumeUp &&
               action.host_action_id == 0;
    case ActionType::KeyboardShortcut:
        return action.key_code != 0 && action.modifiers != 0 &&
               (action.modifiers & ~actions::VALID_MODIFIERS) == 0 &&
               action.consumer == actions::ConsumerAction::VolumeUp &&
               action.host_action_id == 0;
    case ActionType::ConsumerControl:
        return action.key_code == 0 && action.modifiers == 0 &&
               static_cast<uint8_t>(action.consumer) <=
                   static_cast<uint8_t>(actions::ConsumerAction::PlayPause) &&
               action.host_action_id == 0;
    case ActionType::HostAction:
        return action.key_code == 0 && action.modifiers == 0 &&
               action.consumer == actions::ConsumerAction::VolumeUp;
    }
    return false;
}

bool decode_action(const uint8_t* data, std::size_t length,
                   actions::Action& action) {
    if (data == nullptr || length != ACTION_PAYLOAD_SIZE) return false;
    action.type = static_cast<actions::ActionType>(data[0]);
    action.key_code = data[1];
    action.modifiers = data[2];
    action.consumer = static_cast<actions::ConsumerAction>(data[3]);
    action.host_action_id = read_u32(&data[4]);
    return is_valid_action(action);
}

void encode_action(const actions::Action& action, uint8_t* data) {
    data[0] = static_cast<uint8_t>(action.type);
    data[1] = action.key_code;
    data[2] = action.modifiers;
    data[3] = static_cast<uint8_t>(action.consumer);
    write_u32(&data[4], action.host_action_id);
}
}

File: software/StreamDeckDIY.Protocol/ProtocolV1/ActionCodec.cs
using System.Buffers.Binary;
using StreamDeckDIY.Protocol.Models;

namespace StreamDeckDIY.Protocol.ProtocolV1;

public static class ActionCodec
{
    public const int SerializedSize = 8;
    private const ActionModifiers ValidModifiers =
        ActionModifiers.Control | ActionModifiers.Shift |
        ActionModifiers.Alt | ActionModifiers.Gui;

    public static byte[] Serialize(DeviceAction action)
    {
        Validate(action);
        var data = new byte[SerializedSize];
        data[0] = (byte)action.Type;
        data[1] = action.KeyCode;
        data[2] = (byte)action.Modifiers;
        data[3] = (byte)action.ConsumerControl;
        BinaryPrimitives.WriteUInt32LittleEndian(data.AsSpan(4, 4), action.HostActionId);
        return data;
    }

    public static DeviceAction Deserialize(ReadOnlySpan<byte> data)
    {
        if (data.Length != SerializedSize)
        {
            throw new ProtocolException($"Action must contain {SerializedSize} bytes.");
        }

        var action = new DeviceAction(
            (ActionType)data[0],
            data[1],
            (ActionModifiers)data[2],
            (ConsumerControlAction)data[3],
            BinaryPrimitives.ReadUInt32LittleEndian(data.Slice(4, 4)));
        Validate(action);
        return action;
    }

    public static void Validate(DeviceAction action)
    {
        var valid = action.Type switch
        {
            ActionType.None =>
                action.KeyCode == 0 && action.Modifiers == ActionModifiers.None &&
                action.ConsumerControl == ConsumerControlAction.VolumeUp &&
                action.HostActionId == 0,
            ActionType.Keyboard =>
                action.KeyCode != 0 && action.Modifiers == ActionModifiers.None &&
                action.ConsumerControl == ConsumerControlAction.VolumeUp &&
                action.HostActionId == 0,
            ActionType.KeyboardShortcut =>
                action.KeyCode != 0 && action.Modifiers != ActionModifiers.None &&
                (action.Modifiers & ~ValidModifiers) == 0 &&
                action.ConsumerControl == ConsumerControlAction.VolumeUp &&
                action.HostActionId == 0,
            ActionType.ConsumerControl =>
                action.KeyCode == 0 && action.Modifiers == ActionModifiers.None &&
                action.ConsumerControl <= ConsumerControlAction.PlayPause &&
                action.HostActionId == 0,
            ActionType.HostAction =>
                action.KeyCode == 0 && action.Modifiers == ActionModifiers.None &&
                action.ConsumerControl == ConsumerControlAction.VolumeUp,
            _ => false,
        };

        if (!valid)
        {
            throw new ProtocolException("Action contains an invalid field combination.");
        }
    }
}
